code/go/0chain.net/smartcontract/dbs/event/provider.go

```diff
@@ -35,3 +54,69 @@ func (p *Provider) BeforeCreate(tx *gorm.DB) (err error) {
 	}
 	return
 }
+
+// ProviderHealthCheck Update given provider (providerType, providerId) with the given HealthCheckData and calculate aggregates e.g. downtime
+func (edb *EventDb) ProviderHealthCheck(providerType spenum.Provider, providerId string, data interface{}) (err error) {
+	providerModel, err	  := edb.getProviderModel(providerType)
+	if err != nil {
+		return fmt.Errorf("provider_health_check: %v", err.Error()) 
+	}
+	providerInstance, err := edb.getProviderInstance(providerModel, providerId)
+	if err != nil {
+		return fmt.Errorf("provider_health_check: %v", err.Error())
+	}
+
+	healthCheckData, ok := data.(*dbs.DbHealthCheck)
+	if !ok  {
+		return fmt.Errorf("provider_health_check: invalid data")
+	}
+	prevHealthCheck := providerInstance.LastHealthCheck
+	curHealthCheck 	:= healthCheckData.LastHealthCheck
+	diff 			:= curHealthCheck - prevHealthCheck
+	updates 		:= &Provider{}
+	
+	if  diff > (HealthCheckPeriod + HealthCheckDelayLimit) {
+		updates.Downtime = providerInstance.Downtime + uint64(diff)
+	}
+	updates.LastHealthCheck = curHealthCheck
+
+	if err = edb.updateProvider(providerModel, providerId, updates); err != nil {
+		return fmt.Errorf("provider_health_check: %v", err.Error())
+	}
+	return nil
+}
+
+// getProviderModel - Given provider type enum, return a model representing the type of this provider e.g. MinerType => *event.Miner{}
+func (edb *EventDb) getProviderModel(providerType spenum.Provider) (interface{}, error) {
+	providerModel, ok := ProviderModel[providerType]
+	if !ok {
+		return nil, fmt.Errorf("invalid provider type %v", providerType)
+	}
+
+	return providerModel, nil
+}
+
+// getProviderInstance - Given provider model and id, find the corresponding entry to this provider in events_db
+func (edb *EventDb) getProviderInstance(providerModel interface{}, providerId string) (*Provider, error) {
+	providerInstance := Provider{}
+	result := edb.Get().
+		Model(providerModel).
+		Where("id = ?", providerId).
+		Find(&providerInstance)
+	if result.Error != nil {
+		return nil, fmt.Errorf("cannot fetch provider from events_db, providerType = %v, providerId = %v", reflect.TypeOf(providerModel), providerId)
+	}
+	return &providerInstance, nil
+}
+
+// updateProviderInstance - Given provider model and id, update provider-related date in events_db given a payload
+func (edb *EventDb) updateProvider(providerModel interface{}, providerId string, providerUpdates *Provider) error {
```

Comment:
why don't we use event merging here, I believe we should stick to existent framework updating db with events, check https://github.com/0chain/0chain/blob/ce8db27011d6f52536a49c692b94fbd030639c7f/code/go/0chain.net/smartcontract/dbs/event/blobber.go#L197

Reply:
I saw this in the other events and in `process.go` however I didn't investigate to see where it's actually used and if I'm going to need it or not. Now added to specific provider files (miner.go, sharder.go, ...).

Reply:
Please merge events and process them properly.